### ISSUE TITLE
SSIM Evaluation Implementation Question

### ISSUE THREAD

**Issue by triton99** [snap-research/R2L#17]
Hi, thanks for your exciting work.

I successfully reproduced your paper's results on Blender datasets using your provided checkpoints.
I observed that your SSIM metrics are significantly higher than other existing works, ~ 0.9939 on average.

However, I found that there is a weird implementation in your _ssim function. In this function https://github.com/snap-research/R2L/blob/be28bbce3f434fa059039ad9c902dc7fc151ddc9/utils/ssim_torch.py#L87C1-L87C37, the img1 is N,C,H,W, but the actual input is in the N,H,W,C format. I added a permute code as img1 = img1.permute(0,3,1,2) to make it N,C,H,W and I got the results as the below figure.
![image](https://github.com/snap-research/R2L/assets/53249939/e23bd9d3-2f56-45b5-8919-2dc37cf7e838)

For more evidence, I try to compute the ssim of 2 random tensors with your original ssim and the modified one. I got 0.52 with your code and ~0.01 with the modification. It does not make sense that two random tensors have 0.52 in ssim.
Are there any chances that this is an implementation mistake? This issue is serious because the numbers were written in your official paper. And Can you provide the checkpoint for LLFF or the numbers for verification?

**Comment by JuanLuisGonzalez** [comment]
I am also trying to cite and compare against R2L, but I cannot make sense of the SSIM metric. Seems to be too high and not corresponding to the PSNR metric.

**Comment by MingSun-Tse** [comment]
Hi @triton99 , thanks for reporting this problem - short answer: Yes, this _is_ an SSIM implementation mistake.

The implementation I used, as documented in [the header of the code](https://github.com/snap-research/R2L/blob/be28bbce3f434fa059039ad9c902dc7fc151ddc9/utils/ssim_torch.py#L87C1-L87C37), is from another paper. Not noted that their implementation actually used [N, C, H, W] while I used [N, H, W, C]. This is my mistake (sorry...)

Your modified results in the table above probably are the correct ones. I'll do my best to update these numbers as soon as I can. 

For anyone who needs to compare with R2L, just be aware that the **SSIM numbers reported in the paper are incorrect**. Before I can update the paper, one way for you to do the comparison is to run our released checkpoints with the test set, and then _use your own SSIM calculation code_ (I suppose yours is correct).

Thank you guys for the feedback! Regards,

---
07/29 Update:
Here are the new SSIM results I got with the corrected code:
<img width="1286" alt="image" src="https://github.com/snap-research/R2L/assets/16812048/f677646f-3b4e-41fb-87f3-ffb53f43b64e">
The corrected code will be synced to the repo soon.
